File: 2024/9/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* read_input(const char* filename) {
    FILE* file = fopen(filename, "r");
    if (!file) {
        perror("Error opening file");
        exit(1);
    }

    fseek(file, 0, SEEK_END);
    long size = ftell(file);
    rewind(file);

    char* data = (char*)malloc(size + 1);
    if (!data) {
        perror("Memory allocation failed");
        exit(1);
    }

    fread(data, 1, size, file);
    data[size] = '\0';

    fclose(file);
    return data;
}

void input_parse(const char* input, long long** sequence, long long** location, long long** length, long long* seq_len) {
    long long file_id = 0;
    int is_file = 1;
    long long seq_pos = 0;
    long long seq_capacity = 1024;

    *sequence = (long long*)malloc(seq_capacity * sizeof(long long));
    *location = (long long*)malloc(seq_capacity * sizeof(long long));
    *length = (long long*)malloc(seq_capacity * sizeof(long long));

    if (!(*sequence) || !(*location) || !(*length)) {
        perror("Memory allocation failed");
        exit(1);
    }

    for (long long i = 0; i < strlen(input); i++) {
        char ch = input[i];

        if (is_file) {
            (*location)[file_id] = seq_pos;
            (*length)[file_id] = ch - '0';
            is_file = 0;
        } else {
            is_file = 1;
        }

        for (long long j = 0; j < (ch - '0'); j++) {
            if (seq_pos >= seq_capacity) {
                seq_capacity *= 2;
                *sequence = realloc(*sequence, seq_capacity * sizeof(long long));
                *location = realloc(*location, seq_capacity * sizeof(long long));
                *length = realloc(*length, seq_capacity * sizeof(long long));
            }
            (*sequence)[seq_pos++] = (i % 2 == 0) ? file_id : -1;
        }

        if (i % 2 == 0) {
            file_id++;
        }
    }
    *seq_len = seq_pos;
}

void move_files(long long* sequence, long long* seq_len) {
    long long new_length = 0;

    for (long long i = 0; i < *seq_len; i++) {
        if (sequence[i] != -1) {
            new_length++;
        }
    }

    long long* new_seq = (long long*)malloc(new_length * sizeof(long long));
    if (!new_seq) {
        perror("Memory allocation failed");
        exit(1);
    }

    long long new_index = 0;
    
    for (long long pos = 0; pos < new_length; pos++) {
        if (sequence[pos] == -1) {
            for (long long rpos = *seq_len - 1; rpos >= 0; rpos--) {
                if (sequence[rpos] != -1) {
                    new_seq[new_index++] = sequence[rpos];
                    sequence[rpos] = -1; 
                    break;
                }
            }
        } else {
            new_seq[new_index++] = sequence[pos];
        }
    }

    memcpy(sequence, new_seq, new_length * sizeof(long long));

    *seq_len = new_length;

    free(new_seq);
}

void move_files2(long long* sequence, long long* location, long long* length, long long* seq_len) {
    long long current_file = 0;

    while (length[current_file] > 0) {
        current_file++;
    }
    current_file--;

    for (long long to_move = current_file; to_move >= 0; to_move--) {
        long long free_space = 0;
        long long first_free = 0;

        while (first_free < location[to_move] && free_space < length[to_move]) {
            first_free = first_free + free_space;
            free_space = 0;

            while (sequence[first_free] != -1) {
                first_free++;
            }

            while (first_free + free_space < *seq_len && sequence[first_free + free_space] == -1) {
                free_space++;
            }
        }

        if (first_free >= location[to_move]) {
            continue;
        }

        for (long long i = first_free; i < first_free + length[to_move]; i++) {
            sequence[i] = to_move;
        }

        for (long long i = location[to_move]; i < location[to_move] + length[to_move]; i++) {
            sequence[i] = -1;
        }
    }
}

long long calculate_checksum(long long* sequence, long long seq_len) {
    long long checksum = 0;
    for (long long i = 0; i < seq_len; i++) {
        if (sequence[i] != -1) {
            checksum += i * sequence[i];
        }
    }
    return checksum;
}

int main() {
    printf("\nHello, world!\n");
    printf("Here's the solution to Advent of Code Day 9\n");

    char* input = read_input("input.txt");

    long long* sequence;
    long long* location;
    long long* length;
    long long seq_len;

    input_parse(input, &sequence, &location, &length, &seq_len);
    move_files(sequence, &seq_len);
    long long checksum1 = calculate_checksum(sequence, seq_len);

    input_parse(input, &sequence, &location, &length, &seq_len);
    move_files2(sequence, location, length, &seq_len);
    long long checksum2 = calculate_checksum(sequence, seq_len);

    printf("Part 1 result: %lld\n", checksum1);
    printf("Part 2 result: %lld\n", checksum2);

    free(input);
    free(sequence);
    free(location);
    free(length);

    return 0;
}

File: 2024/9/main.py
def read_input(filename):
    with open(filename, 'r') as file:
        data = file.read()
    return data

def input_parse(input):
    file_id = 0
    is_file = True
    
    sequence = []
    location, length = [0] * len(input ), [0] * len(input)

    for pos, char in enumerate(input):
        if is_file:
            location[file_id] = len(sequence)
            length[file_id] = int(char)
            is_file = False
        else:
            is_file = True

        for _ in range(int(char)):
            sequence.append(str(file_id) if pos % 2 == 0 else '.')

        if pos % 2 == 0:
            file_id += 1

    return sequence, location, length

def move_files(sequence):
    new_seq = []
    new_length = len(sequence) - sequence.count('.')

    for pos in range(new_length):
        if sequence[pos] == '.':
            for rpos in range(len(sequence)-1, 0, -1):
                if sequence[rpos]!= '.':
                    new_seq.append(sequence[rpos])
                    sequence[rpos] = '.'
                    break
        else:
            new_seq.append(sequence[pos])   

    return new_seq

def move_files2(sequence_list):
    sequence, location, length = sequence_list

    current_file = 0
    while length[current_file] > 0:
        current_file += 1
    current_file -= 1

    for to_move in range(current_file, -1, -1):
        free_space = 0
        first_free = 0

        while first_free < location[to_move] and free_space < length[to_move]:
            first_free = first_free + free_space
            free_space = 0
            while sequence[first_free] != '.':
                first_free += 1
            while first_free + free_space < len(sequence) and sequence[first_free + free_space] == '.':
                free_space += 1

        if first_free >= location[to_move]:
            continue

        for i in range(first_free, first_free + length[to_move]):
            sequence[i] = to_move
        for i in range(location[to_move], location[to_move] + length[to_move]):
            sequence[i] = '.'

    return sequence

def calculate_checksum(sequence):
    checksum = 0

    for i, file in enumerate(sequence):
        if file != '.':
            checksum += i * int(file)

    return checksum

def main():
    print("\nHello, world!\n")
    print("Here's the solution to Advent of Code Day 9\n")

    input = read_input('input.txt')
    
    files = move_files(input_parse(input)[0])
    files2 = move_files2(input_parse(input))

    print("Part 1 result : ", calculate_checksum(files))
    print("Part 2 result : ", calculate_checksum(files2))
    
if __name__ == '__main__':
    main()